device/src/region/fixed_channel_plans/mod.rs

```diff
@@ -2,21 +2,170 @@ use super::*;
 use core::marker::PhantomData;
 use lorawan::maccommands::ChannelMask;
 
+// Compiler trick needed to assert array length at compile time
+#[macro_export]
+macro_rules! gen_assert {
+    ($t:ident, $c:expr) => {{
+        struct Check<const $t: usize>(usize);
+        impl<const $t: usize> Check<$t> {
+            const CHECK: () = assert!($c);
+        }
+        let _ = Check::<$t>::CHECK;
+    }};
+}
+
 mod au915;
 mod us915;
 
 pub(crate) use au915::AU915;
 pub(crate) use us915::US915;
 
+seq_macro::seq!(
+    N in 0..=71 {
+        #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+        #[repr(u8)]
+        pub enum Channel {
+            #(
+                _~N,
+            )*
+        }
+    }
+);
+
+impl From<Channel> for u8 {
+    fn from(value: Channel) -> Self {
+        value as u8
+    }
+}
+
+#[derive(Clone)]
+struct PreferredJoinChannels {
```

Comment:
You'll notice if you build with all features enabled, we have an issue since this struct doesn't implement Serialize/Deserialize. If you keep digging in, you'll see heapless::Vec doesn't implement Serialize/Deserialize.... so we'd have to figure that out if we maintain this heapless::Vec of preferred channels. We could of course write a Serialize/Deserialize implementation...

However, I'm curious what this looks like if "preferred channels" are described in a `ChannelMask<9>`. We already have functions to enable/disable channels and subbands on that struct. And we know how to derive valid channels from that struct and after `max_num_retries`, we could reset the channel mask to default (ie: all enabled).

And just to be clear, I'm not saying that that approach is "the way". I just have a feeling things will look simpler with it. I'll maybe try to PR against your branch so we can see how it looks comparatively.

Reply:
Yep. I'll see what it looks like. Take all of the latest changes with a grain of salt, the PR isn't finished. I just wanted your input on the compile time checking side of things before I kept going.